tools/workflow/cleaner/apigw/clean.go

```diff
@@ -0,0 +1,80 @@
+package apigw
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/apigateway"
+	"log"
+	"os"
+	"regexp"
+	"time"
+)
+
+const Type = "apigw"
+
+var logger = log.New(os.Stdout, fmt.Sprintf("[%s] ", Type), log.LstdFlags)
+
+func Clean(sess *session.Session, expirationDate time.Time) error {
+	logger.Printf("Begin to clean APIGW clients")
+
+	apigwclient := apigateway.New(sess)
+	var deleteGatewayIds []*string
+	var nextToken *string
+	for {
+		getRestApisInput := &apigateway.GetRestApisInput{Position: nextToken}
+		getRestApisOutput, err := apigwclient.GetRestApis(getRestApisInput)
+		if err != nil {
+			return fmt.Errorf("unable to retrieve APIs: %w", err)
+		}
+		for _, gw := range getRestApisOutput.Items {
+			doesNameMatch, err := shouldDelete(gw)
+			if err != nil {
+				return fmt.Errorf("error during pattern match: %w", err)
+			}
+			if expirationDate.After(*gw.CreatedDate) && doesNameMatch {
+				logger.Printf("Try to delete gateway %s created-at %v", *gw.Id, gw.CreatedDate)
+				deleteGatewayIds = append(deleteGatewayIds, gw.Id)
+			}
+		}
+		if getRestApisOutput.Position == nil {
+			break
+		}
+		nextToken = getRestApisOutput.Position
+
+		if len(deleteGatewayIds) < 1 {
```

Comment:
Is this right?  What if the first page of results produced nothing to delete but subsequent pages would?

It feels like the `for` block containing this should end just before this line and the function from here out executed once and only once when the full list of gateways to delete has been constructed.